bon-macros/src/builder/builder_gen/build_from.rs

```diff
@@ -0,0 +1,133 @@
+use proc_macro2::Span;
+use quote::quote;
+use syn::{Type, spanned::Spanned};
+
+pub(super) fn emit(ctx: &BuilderGenCtx, target_ty: &Type) -> Result<TokenStream> {
+    let mut tokens = TokenStream::new();
+
+    let field_vars: Vec<_> = ctx
+        .members
+        .iter()
+        .map(|member| {
+            let ident = member.orig_ident();
+            let ty = member.norm_ty();
+            let default_expr = quote! { ::core::default::Default::default() };
+
+            match member {
+                Member::Field(_) | Member::StartFn(_) => quote! {
+                    let #ident: #ty = self.#ident;
+                },
+                Member::Named(member) => {
+                    let index = &member.index;
+                    quote! {
+                        let #ident: #ty = match self.__unsafe_private_named.#index {
+                            Some(value) => value,
+                            None => from.#ident.clone(),
+                        };
+                    }
+                }
+                Member::FinishFn(_) => quote! {
+                    let #ident: #ty = from.#ident.clone();
+                },
+                Member::Skip(_) => quote! {
+                    let #ident: #ty = #default_expr;
+                },
+            }
+        })
+        .collect();
+
+    let ctor_args: Vec<_> = ctx
+        .members
+        .iter()
+        .map(|m| {
+            let ident = m.orig_ident();
+            quote! { #ident }
+        })
+        .collect();
+
+    if ctx.build_from {
+        tokens.extend(emit_build_from_method(
+            false,
+            target_ty,
+            &field_vars,
+            &ctor_args,
+        ));
+    }
+
+    if ctx.build_from_clone {
+        tokens.extend(emit_build_from_method(
+            true,
+            target_ty,
+            &field_vars,
+            &ctor_args,
+        )?);
+    }
+
+    Ok(tokens)
+}
+
+fn emit_build_from_method(
+    clone: bool,
+    target_ty: &Type,
+    field_vars: &[TokenStream],
+    ctor_args: &[TokenStream],
+) -> Result<TokenStream> {
+    let doc = if clone {
+        "Fills unset builder fields from an owned value of the target type and builds it."
+    } else {
+        "Fills unset builder fields from a reference to the target type and builds it."
+    };
+
+    let method_name = if clone {
+        quote!(build_from_clone)
+    } else {
+        quote!(build_from)
+    };
+
+    let arg_type = if clone {
+        quote!(#target_ty)
+    } else {
+        quote!(&#target_ty)
+    };
+
+    let arg_pat = if clone {
+        quote!(mut from)
+    } else {
+        quote!(from)
+    };
+
+    let ctor_path = extract_ctor_ident_path(target_ty, target_ty.span())?;
+
+    Ok(quote! {
+        #[inline(always)]
+        #[doc = #doc]
+        pub fn #method_name(self, #arg_pat: #arg_type) -> #target_ty {
+            #( #field_vars )*
+            #ctor_path {
+                #( #ctor_args, )*
+            }
```

Comment:
Unfortunatelly, this approach won't work in case if the builder is generated with the `#[builder]` macro on top of a function. We have the `body.generate()` in `finish_fn.rs` here that generates the proper body: https://github.com/elastio/bon/blob/af6e9251ec65217e974ec3dcc447f9c1e456cd72/bon-macros/src/builder/builder_gen/finish_fn.rs#L139

It's only requirement is that members are assigned to variables of the same name in scope. Plus we need to inherit the potential `async` / `unsafe` / `const` modifiers and maybe handle the case of the `Result` from the original function